src/EventStore.Core/TransactionLog/Scavenging/TransactionManager.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System;
+
+namespace EventStore.Core.TransactionLog.Scavenging {
+	// This makes sure we dont accidentally start trying to nest transactions or begin them concurrently
+	// and facilities commiting the open transaction with a checkpoint
+	public class TransactionManager<TTransaction> : ITransactionManager {
+		private readonly ITransactionFactory<TTransaction> _factory;
+		private readonly IScavengeMap<Unit, ScavengeCheckpoint> _storage;
+		private Action _onRollback;
+		private bool _began;
+		private TTransaction _transaction;
+
+		public TransactionManager(
+			ITransactionFactory<TTransaction> factory,
+			IScavengeMap<Unit, ScavengeCheckpoint> storage) {
+
+			_factory = factory;
+			_storage = storage;
+		}
+
+		public void RegisterOnRollback(Action onRollback) {
+			if (_onRollback != null)
+				throw new InvalidOperationException();
+
+			_onRollback = onRollback;
+		}
+
+		public void UnregisterOnRollback() {
+			_onRollback = null;
+		}
+
+		public void Begin() {
+			if (_began)
+				throw new InvalidOperationException("Cannot begin a transaction that has already begun.");
+
+			_transaction = _factory.Begin();
+			_began = true;
+		}
+
+		public void Rollback() {
+			if (!_began)
+				throw new InvalidOperationException("Cannot rollback a transaction that has not begun.");
+
+			_factory.Rollback(_transaction);
+			_onRollback?.Invoke();
+			_began = false;
+		}
+
+		public void Commit(ScavengeCheckpoint checkpoint) {
+			if (!_began)
+				throw new InvalidOperationException("Cannot commit a transaction that has not begun.");
+
+			_storage[Unit.Instance] = checkpoint;
+
+			_factory.Commit(_transaction);
```

Comment:
is the order correct? shouldn't the checkpoint be set after doing the commit?

Reply:
oh i see, the checkpoint is also committed as part of the same transaction?  :thinking:

Reply:
correct, the checkpoint is committed as part of the transaction

Reply:
confirmed by Tim: the checkpoint is committed as part of the transaction